modules/bgsegm/src/bgfg_lsbp.cpp

```diff
@@ -0,0 +1,698 @@
+
+public:
+    BackgroundSubtractorLSBPImpl(int mc,
+                                 int nSamples,
+                                 float replaceRate,
+                                 float propagationRate,
+                                 int hitsThreshold,
+                                 float alpha,
+                                 float beta,
+                                 float blinkingSupressionDecay,
+                                 float blinkingSupressionMultiplier,
+                                 float noiseRemovalThresholdFacBG,
+                                 float noiseRemovalThresholdFacFG,
+                                 bool useDescriptors);
+
+    CV_WRAP virtual void apply(InputArray image, OutputArray fgmask, double learningRate = -1);
+
+    CV_WRAP virtual void getBackgroundImage(OutputArray backgroundImage) const;
+
+    friend class ParallelLSBP;
+};
+
+class ParallelLSBP : public ParallelLoopBody {
+private:
+    const Size sz;
+    BackgroundSubtractorLSBPImpl* bgs;
+    const Mat& frame;
+    const double learningRate;
+    const Mat& LSBPDesc;
+    Mat& fgMask;
+
+    ParallelLSBP &operator=(const ParallelLSBP&);
+
+public:
+    ParallelLSBP(const Size& _sz, BackgroundSubtractorLSBPImpl* _bgs, const Mat& _frame, double _learningRate, const Mat& _LSBPDesc, Mat& _fgMask)
+    : sz(_sz), bgs(_bgs), frame(_frame), learningRate(_learningRate), LSBPDesc(_LSBPDesc), fgMask(_fgMask) {};
+
+    void operator()(const Range &range) const {
+        BackgroundModel* backgroundModel = bgs->backgroundModel.get();
+        Mat& distMovingAvg = bgs->distMovingAvg;
+
+        for (int index = range.start; index < range.end; ++index) {
+            const int i = index / sz.width, j = index % sz.width;
+            int k;
+            const float minDist = backgroundModel->findClosest(i, j, frame.at<Point3f>(i, j), k);
+
+            distMovingAvg.at<float>(i, j) *= 1 - float(learningRate);
+            distMovingAvg.at<float>(i, j) += float(learningRate) * minDist;
+
+            const float threshold = bgs->alpha * distMovingAvg.at<float>(i, j) + bgs->beta;
+            BackgroundSample& sample = (* backgroundModel)(k);
+
+            if (LSBPDist32(sample.desc ^ LSBPDesc.at<uint32_t>(i, j)) > LSBPthreshold || minDist > threshold) {
+                fgMask.at<uint8_t>(i, j) = 255;
+
+                if (bgs->rng.uniform(0.0f, 1.0f) < bgs->replaceRate)
+                    backgroundModel->replaceOldest(i, j, BackgroundSample(frame.at<Point3f>(i, j), LSBPDesc.at<uint32_t>(i, j), bgs->currentTime));
+            }
+            else {
+                sample.color *= 1 - learningRate;
+                sample.color += learningRate * frame.at<Point3f>(i, j);
+                sample.time = bgs->currentTime;
+                ++sample.hits;
+
+                // Propagation to neighbors
+                if (sample.hits > bgs->hitsThreshold && bgs->rng.uniform(0.0f, 1.0f) < bgs->propagationRate) {
```

Comment:
If I understand correctly, only one randomly selected 8-neighboring sample should be updated.